Reintroduce TypeScript docs (#2153)
wasp-lang/wasp: web/docs/general/typescript.md

Author: Repurposed the old TypeScript page for a short migration guide. Most of the text is old.

---
title: TypeScript Support
---

import TypescriptServerNote from '../_TypescriptServerNote.md'

# TypeScript support

> TypeScript is a programming language that brings static type analysis to JavaScript. It is a superset of JavaScript (i.e., all valid JavaScript programs are valid TypeScript programs) and compiles to JavaScript before running.
> 
> TypeScript's type system detects common errors at build time (reducing the chance of runtime errors in production) and enables type-based auto-completion in IDEs.
> 
> Each Wasp feature comes with accompanying TypeScript documentation.
> 
> If you're starting a new project and want to use TypeScript, you don't need to do anything special.
> Just follow the feature docs you are interested in, and they will tell you everything you need to know.
> We recommend you start by going through [the tutorial](../tutorial/01-create.md).
> 
> If you already have a Wasp project written in JavaScript and you wish to
> migrate it to TypeScript, then this guide is meant for you.

Reviewer: Some language simplification:
```suggestion
TypeScript is a programming language that adds static type analysis to JavaScript. It is a superset of JavaScript, meaning all JavaScript code is valid TypeScript code, and it compiles to JavaScript.

TypeScript's type system helps catch errors at build time, reducing runtime errors, and provides type-based auto-completion in IDEs.

Each Wasp feature includes TypeScript documentation.

If you're starting a new project and want to use TypeScript, you don't need to do anything special.
Just follow the feature docs you are interested in, and they will tell you everything you need to know.
We recommend you start by going through [the tutorial](../tutorial/01-create.md).

To migrate an existing Wasp project from JavaScript to TypeScript, follow this guide.
```

I'd definitely go for more basic English since I suspect it's easier to read through (especially non-native speakers).

## Migrating your project to TypeScript
Wasp supports TypeScript out of the box.

Our scaffolding already includes TypeScript, so migrating your project to TypeScript is as simple as changing file extensions and using the language. This approach allows you to gradually migrate your project on a file-by-file basis.

We will first show you how to migrate a single file and then help you generalize the procedure to the rest of your project.

### Migrating a single file

Assuming your `schema.prisma` file defines the `Task` entity:

```prisma title="schema.prisma"
// ...

model Task {
  id          Int @id @default(autoincrement())
  description String
  isDone      Boolean
}
```
And your `main.wasp` file defines the `getTaskInfo` query:

```wasp title=main.wasp
query getTaskInfo {
  fn: import { getTaskInfo } from "@src/queries",
  entities: [Task]
}
```

> 

Author: Does anyone know why our `.prettierrc` ignores `.md` files?

We will show you how to migrate the following `queries.js` file:

```javascript title="src/queriesjs"
import HttpError from "wasp/server"

function getInfoMessage(task) {
  const isDoneText = task.isDone ? "is done" : "is not done"
  return `Task '${task.description}' is ${isDoneText}.`
}

export const getTaskInfo = async ({ id }, context) => {
  const Task = context.entities.Task
  const task = await Task.findUnique({ where: { id } })
  if (!task) {
    throw new HttpError(404)
  }
  return getInfoMessage(task)
}
```

To migrate this file to TypeScript, all you have to do is:

> 1.  Change the filename from `queries.js` to `queries.ts`.

Reviewer: ```suggestion
1. Change the filename from `queries.js` to `queries.ts`.
```

---
title: TypeScript Support
---

import TypescriptServerNote from '../_TypescriptServerNote.md'

# TypeScript support

TypeScript is a programming language that adds static type analysis to JavaScript.
It is a superset of JavaScript, which means all JavaScript code is valid TypeScript code.
It also compiles to JavaScript before running.

TypeScript's type system helps catch errors at build time (this reduces runtime errors), and provides type-based auto-completion in IDEs.

Each Wasp feature includes TypeScript documentation.

If you're starting a new project and want to use TypeScript, you don't need to do anything special.
Just follow the feature docs you are interested in, and they will tell you everything you need to know.
We recommend you start by going through [the tutorial](../tutorial/01-create.md).

To migrate an existing Wasp project from JavaScript to TypeScript, follow this guide.

## Migrating your project to TypeScript

> Wasp supports TypeScript out of the box.

Reviewer: This sentence feels loose, floating alone up here. Maybe remove it or join it with the text right below it.

Author: I tried moving it down, but it looked weirder:
![image](https://github.com/user-attachments/assets/e2c4bb32-ac68-4c52-b903-4f75a341784c)

So I changed the text to this:
![image](https://github.com/user-attachments/assets/94ec45e5-f203-464e-b5b3-8ddb8574c0de)

Our scaffolding already includes TypeScript, so migrating your project to TypeScript is as simple as changing file extensions and using the language. This approach allows you to gradually migrate your project on a file-by-file basis.

We will first show you how to migrate a single file and then help you generalize the procedure to the rest of your project.

### Migrating a single file

Assuming your `schema.prisma` file defines the `Task` entity:

```prisma title="schema.prisma"
// ...

model Task {
  id          Int @id @default(autoincrement())
  description String
  isDone      Boolean
}
```

And your `main.wasp` file defines the `getTaskInfo` query:

```wasp title=main.wasp
query getTaskInfo {
  fn: import { getTaskInfo } from "@src/queries",
  entities: [Task]
}
```

We will show you how to migrate the following `queries.js` file:

> ```javascript title="src/queriesjs"

Reviewer: ```suggestion
```javascript title="src/queries.js"
```

import HttpError from 'wasp/server'

function getInfoMessage(task) {
  const isDoneText = task.isDone ? 'is done' : 'is not done'
  return `Task '${task.description}' is ${isDoneText}.`
}

export const getTaskInfo = async ({ id }, context) => {
  const Task = context.entities.Task
  const task = await Task.findUnique({ where: { id } })
  if (!task) {
    throw new HttpError(404)
  }
  return getInfoMessage(task)
}
```

To migrate this file to TypeScript, all you have to do is:

1. Change the filename from `queries.js` to `queries.ts`.
2. Write some types (and optionally use some of Wasp's TypeScript features).

<Tabs>
<TabItem value="before" label="Before">

> ```javascript title="src/queries.ts"

Reviewer: ```suggestion
```javascript title="src/queries.js"
```

import HttpError from '@wasp/core/HttpError.js'

function getInfoMessage(task) {
  const isDoneText = task.isDone ? 'is done' : 'is not done'
  return `Task '${task.description}' is ${isDoneText}.`
}

export const getTaskInfo = async ({ id }, context) => {
  const Task = context.entities.Task
  const task = await Task.findUnique({ where: { id } })
  if (!task) {
    throw new HttpError(404)
  }
  return getInfoMessage(task)
}
```

</TabItem>
<TabItem value="after" label="After">

> ```typescript title=src/server/queries.ts

Reviewer: ```suggestion
```typescript title=src/queries.ts
```
